dashboard/lib/email_reminder.rb

```diff
@@ -0,0 +1,148 @@
+require 'stringio'
+require 'cdo/aws/metrics'
+require 'cdo/aws/s3'
+require 'cdo/chat_client'
+
+class EmailReminder
+  include Rails.application.routes.url_helpers
+  attr_reader :max_reminder_age, :min_reminder_age, :dry_run, :log
+
+  # Max reminders we want to send over the lifetime of a permission request
+  # Currently, we only want to send 1 reminder to a parent for CPA compliance.
+  MAX_LIFETIME_REMINDERS = 1
+
+  def initialize(options = {})
+    # Oldest requests we want to send reminders for. Defaults to 7 days ago.
+    @max_reminder_age = options[:max_reminder_age] || 7.days.ago
+    raise ArgumentError.new('max_reminder_age must be Time') unless @max_reminder_age.is_a? Time
+
+    # Newest requests we want to send reminders for. Defaults to 3 days ago.
+    @min_reminder_age = options[:min_reminder_age] || 3.days.ago
+    raise ArgumentError.new('min_reminder_age must be Time') unless @min_reminder_age.is_a? Time
+    raise ArgumentError.new('max_reminder_age must come before min_reminder_age') unless @max_reminder_age < @min_reminder_age
+
+    @dry_run = options[:dry_run].nil? ? false : options[:dry_run]
+    raise ArgumentError.new('dry_run must be boolean') unless [true, false].include? @dry_run
+
+    reset
+    @log.puts "Initialized with options: #{options.inspect}"
```

Comment:
Are there any formal specs on the logging feature in this class? Do we need to maintain an audit log of the email reminders we’ve sent to parents as part of compliance with this law?

The reason I ask is that if this is just standard application logging to aid in debugging and monitoring, we have an undocumented standard of using `CDO.log.info` which streams to AWS CloudWatch Logs. Here’s an example.

Reply:
I don't believe there's a compliance need for auditing reminders. The reminders are more of a nice-to-have feature.

The logging was just following the pattern of other cronjobs, but switching to CDO.log works for me if that's the standard.

Reply:
ah … I’m seeing that some other scheduled jobs created by Platform team (Contact Rolllups 2.0, Deleted Accounts Purger, etc.) have been logging to the bucket `cdo-audit-logs`. I think from an Infrastructure perspective, we would recommend using `CDO.log.info` instead in an effort to standardize our logging Sorry this is not formally agreed upon or documented. 

At least one technical advantage is that `CDO.log.info` will write somewhat synchronously to AWS CloudWatch Logs each time it is invoked. Uploading a single log file at the end of a job’s execution is asynchronous so there’s less information available to debug a long-running or stuck job. Also, if we don’t upload the log in an `ensure` block an error raised during execution might prevent any logging from occurring.

Reply:
Ok, I'll look into `CDO.log.info`. This might also help me out with an error that I can't repro locally where the tests are calling out to the EC2 instance metadata service somewhere in the code that does the s3 upload, even though I've stubbed it out.